### ISSUE TITLE
Valgrind complains about uninitialized values

### ISSUE THREAD

**Issue by xTibor** [chyh1990/yaml-rust#161]
I noticed while debugging some Rust apps this crate was throwing uninitialized value warnings in valgrind. I don't know if this is real UB or bogus warnings, just reporting it to be safe.

**Rust version:**
`rustc 1.43.1 (8d69840ab 2020-05-04)`

**Test case:**
https://gist.github.com/xTibor/8858ecf5cbd276bc8673267ee5d324ac

**Valgrind logs:**
```
==11197== Conditional jump or move depends on uninitialised value(s)
==11197==    at 0x1252C3: yaml_rust::parser::Parser<T>::parser_process_directives (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x1250F3: yaml_rust::parser::Parser<T>::document_start (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x122112: yaml_rust::parser::Parser<T>::next (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x12360A: yaml_rust::parser::Parser<T>::load (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x11E247: yaml_rust::yaml::YamlLoader::load_from_str (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x10D892: overrides::main (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x10D4B2: std::rt::lang_start::{{closure}} (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==    by 0x1362B2: {{closure}} (rt.rs:52)
==11197==    by 0x1362B2: std::panicking::try::do_call (panicking.rs:303)
==11197==    by 0x137FF6: __rust_maybe_catch_panic (lib.rs:86)
==11197==    by 0x136CBB: try<i32,closure-0> (panicking.rs:281)
==11197==    by 0x136CBB: catch_unwind<closure-0,i32> (panic.rs:394)
==11197==    by 0x136CBB: std::rt::lang_start_internal (rt.rs:51)
==11197==    by 0x10DA31: main (in /home/user/git/yaml-rust/target/release/examples/overrides)
==11197==  Uninitialised value was created by a stack allocation
==11197==    at 0x125250: yaml_rust::parser::Parser<T>::parser_process_directives (in /home/user/git/yaml-rust/target/release/examples/overrides)
```

**Comment by dtolnay** [comment]
This crate contains zero unsafe code so the root cause is almost surely not in this crate.

**Comment by hoodie** [comment]
linked hashmap does though

**Comment by dtolnay** [comment]
Linked-hash-map issues should be reported to linked-hash-map.

**Comment by dtolnay** [comment]
This reproduces without linked-hash-map though, so it looks like an LLVM or rustc bug. Reduced somewhat:

```rust
use std::collections::VecDeque;

struct Parser {
    scanner: VecDeque<Token>,
    token: Option<Token>,
}

enum Token {
    #[allow(dead_code)]
    VersionDirective,
    DocumentStart,
    #[allow(dead_code)]
    Tag(String, String),
}

fn main() {
    let mut parser = Parser {
        scanner: VecDeque::new(),
        token: None,
    };
    peek_token(&mut parser).unwrap();
    let _ = match peek_token(&mut parser).unwrap() {
        Token::VersionDirective => explict_document_start(&mut parser),
        _ => parser_process_directives(&mut parser),
    };
}

fn peek_token(parser: &mut Parser) -> Result<&Token, String> {
    match parser.token {
        None => {
            let token = Token::DocumentStart;
            parser.scanner.push_back(token);
            let token = parser.scanner.pop_front();
            parser.token = Some(token.unwrap());
            Ok(parser.token.as_ref().unwrap())
        }
        Some(ref tok) => Ok(tok),
    }
}

fn skip(parser: &mut Parser) {
    parser.token = None;
}

fn parser_process_directives(parser: &mut Parser) -> Result<(), String> {
    loop {
        match peek_token(parser)? {
            Token::VersionDirective => {}
            _ => break,
        }
        skip(parser);
    }
    Ok(())
}

fn explict_document_start(parser: &mut Parser) -> Result<(), String> {
    parser_process_directives(parser).unwrap();
    match *peek_token(parser)? {
        Token::VersionDirective => {
            skip(parser);
            Ok(())
        }
        _ => Err(String::new()),
    }
}
```

**Comment by dtolnay** [comment]
According to https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/sanitizer.html#memorysanitizer you need to build with `-Zbuild-std` in order for initializations inside the standard library to be understood by MemorySanitizer, otherwise you get false positives. That looks like what this is because building with `-Zbuid-std` does not report any uninitialized accesses.

```console
$ cargo clean
$ export RUSTFLAGS='-Zsanitizer=memory -Zsanitizer-memory-track-origins'
$ cargo run -Zbuild-std --target x86_64-unknown-linux-gnu --release
```
